### PR TITLE
MAINT: stats._axis_nan_policy: preserve dtype of masked arrays where possible

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Followup to gh-15239

#### What does this implement/fix?
When gh-15239 added masked array support to functions with the `_axis_nan_policy` decorator, it did so at the cost of casting inputs to `np.float64`. Consequently, some stats functions no longer performed calculations with the intended dtype when passed masked arrays.

This PR ensures that the conversion is to `np.result_type` of the arrays so that conversions are performed only as needed to keep types consistent.

### CODE REVIEW

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:187]: `else np.float64` was intended to be for types that aren't NumPy numbers but can be converted.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:197]: Originally, I didn't want to have separate logic for different numerical types, so I converted everything to `float64`. I understand the `dtype` hierarchy a bit better now, so I see that we really only need to consider two types - integers and floating point ("inexact") types - and by defining `info` and `nextafter` appropriately, we don't need to branch in the code below.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:203]: No need to re-run CI for this. Change this in next commit or right before merge (with `[skip ci]` comment).
```suggestion
            if np.any(sample == sentinel):  # choose a new sentinel value
```

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:204]: Perhaps I didn't know about `np.nextafter` before? I was multiplying by `(1 - 2*eps)` because I wasn't sure whether `(1 - eps)` would always result in a distinct number. All we need is a distinct number, so `np.nextafter` should be fine for floats and subtracting 1 is fine for integers.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:212]: If we subtract 1 for the lowest integer, we loop to the highest integer. 
Likewise, if we use `np.nextafter` toward `-np.inf` from the minimum float, we get `-np.inf`, and if we do it again we still get `-np.inf`.
Both of these cases could get us into an infinite loop. 
To avoid this, this `else` clause happens before the lowest finite value of the dtype is considered. (See strict greater than in `while sentinel > min_possible`.)

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:218]: We want to convert any masked array to a regular array - even if they have all-zero masks.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:220]: Ideally, this would copy the data if `np.any(mask)` (to avoid modifying it below) but just extract the existing data array otherwise. Unfortunately, the `data` attribute for complex types is a memoryview, so it's not quite as simple as:
```python3
sample = sample.data.copy() if np.any(mask) else sample.data
```
I don't want to dig into this, so just use `np.asarray` to extract the data as a regular array.

File: scipy/stats/tests/test_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/tests/test_axis_nan_policy.py:676]: The algorithm for choosing sentinel values changed a bit, so this needed to be updated.

File: scipy/stats/tests/test_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/tests/test_axis_nan_policy.py:699]: I know this is lengthy, but I thought it would be more readable and require less test execution time if I wrote very specific tests rather than writing shorter but more complex code to test everything under the sun. Hopefully the comments help with interpretation.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py]: Well, turns out I can just do both. One ensure that I don't make a copy if I don't need to, and the other ensures that it's an array rather than a memoryview.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [inline scipy/stats/_axis_nan_policy.py:221]: I don't think `sample.data` can be a memoryview but the docs don't clear this out. Since `np.asarray` should be a no-op if `sample` is already an array, I guess it's ok to keep it.

**Review by tirthasheshpatel — COMMENTED** [review]
The changes look good. Maybe, add a test for `mode` or some other function that preserves dtypes to ensure that the dtype is preserved? Something like:

```python
from scipy import stats
from scipy.stats._stats_py import ModeResult
from scipy.stats._axis_nan_policy import _axis_nan_policy_factory

func = _axis_nan_policy_factory(ModeResult)(stats.mode)
res = func([1,2,3,4])
assert res[0].dtype == np.int64
res = func([['a', 'a', 'b', 'c']], axis=1)
assert np.issubdtype(res[0].dtype, np.str_)
```

File: scipy/stats/tests/test_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/tests/test_axis_nan_policy.py:753]: @tirthasheshpatel added this to respond to your [comment](https://github.com/scipy/scipy/pull/16664#pullrequestreview-1047686381). The decorator still won't work on masked arrays with non-numeric input like strings. (Actually, we held off on adding the decorator to `mode` specifically because we didn't want to deal with non-numeric input.) Also, this is only a problem for masked arrays, so the array has to be masked to test the fix.
Anyway, in main this example would produce `np.float64`; now it correctly produces a `np.float32`.

Comment (tirthasheshpatel) [reply scipy/stats/tests/test_axis_nan_policy.py:753]: Ah, yes. Forgot about the problem with non-numeric arrays. This test is what I had in mind.

**Review by tirthasheshpatel — APPROVED** [review]
LGTM now! Thanks @mdhaber! I will merge once the tests pass.

**Pull request merged**
